### PR TITLE
Add `post_init`

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
Mostly as discussed in the dev chat. Allows users to run custom startup logic.

Also orders `AppBuilder.__slots__` alphabetically

### Checklist for PRs

- [ ] ~Added `.. versionadded:: version`, `.. versionchanged:: version` or `.. deprecated:: version` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)~ Only changed classes that already have the "new in v20" directive"
- [x] Created new or adapted existing unit tests
- [ ] Documented code changes according to the [CSI standard](https://standards.mousepawmedia.com/en/stable/csi.html)
- [ ] Added myself alphabetically to `AUTHORS.rst` (optional)
- [ ] Added new classes & modules to the docs and all suitable `__all__` s

### CODE REVIEW

File: telegram/ext/_application.py [hunk 1]
```diff
@@ -179,6 +179,9 @@ class Application(Generic[BT, CCT, UD, CD, BD, JQ], AbstractAsyncContextManager)
                 :meth:`add_error_handler`
         context_types (:class:`telegram.ext.ContextTypes`): Specifies the types used by this
             dispatcher for the ``context`` argument of handler and job callbacks.
+        post_init (:term:`coroutine function`): Optional. A callback that will be executed by
+            :meth:`Application.run_polling` and :meth:`Application.run_webhook` initializing the
```
Comment (harshil21) [inline telegram/ext/_application.py:183]:
```suggestion
            :meth:`Application.run_polling` and :meth:`Application.run_webhook` when initializing the
```

File: telegram/ext/_applicationbuilder.py [hunk 2]
```diff
@@ -885,6 +899,32 @@ def updater(self: BuilderType, updater: Optional[Updater]) -> BuilderType:
...
+        """
+        Sets a callback to be executed by :meth:`Application.run_polling` and
+        :meth:`Application.run_webhook` *after* executing :meth:`Application.initialize` but
+        *before* executing :meth:`Updater.start_polling` or :meth:`Updater.start_webhook`,
+        respectively.
+
+        Tip:
+            This can be used for custom startup logic that requires to await coroutines, e.g.
+            setting up the bots commands via :meth:`~telegram.Bot.set_my_commands`.
+
```
Comment (harshil21) [inline telegram/ext/_applicationbuilder.py:914]: Maybe worth including an example?
```suggestion
            setting up the bots commands via :meth:`~telegram.Bot.set_my_commands`.
            
            Example::
            
                async def callback(app: Application) -> None:
                    ...
                
                application = Application.builder().token("TOKEN").post_init(callback).build()

```

File: telegram/ext/_applicationbuilder.py [hunk 3]
```diff
@@ -885,6 +899,40 @@ def updater(self: BuilderType, updater: Optional[Updater]) -> BuilderType:
...
+        Example:
+            .. code::
+
+                async def post_init(application: Application) -> None:
```
[the remark this answers is not in the document]
Comment (harshil21) [reply telegram/ext/_applicationbuilder.py:918]: in the builder it's given as callback

[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/_applicationbuilder.py:918]: ah lol xD

**Review by harshil21 — APPROVED** [review]
LGTM

**Pull request merged**